pkg/agent/clean/active_directory.go

```diff
@@ -0,0 +1,972 @@
+			return nil, nil, err
+		}
+	}
+	restConfig.RateLimiter = ratelimit.None
+
+	sc, err := scaledContext(restConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+	adConfig, err := adConfiguration(sc)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return sc, adConfig, nil
+}
+
+// UnmigrateAdGUIDUsersOnce will ensure that the migration script will run only once.  cycle through all users, ctrb, ptrb, tokens and migrate them to an
+// appropriate DN-based PrincipalID.
+func UnmigrateAdGUIDUsersOnce(sc *config.ScaledContext) error {
+	migrationConfigMap, _ := sc.Core.ConfigMaps("cattle-system").GetNamespaced("cattle-system", "ad-guid-migration", metav1.GetOptions{})
+	if migrationConfigMap != nil {
+		migrationStatus := migrationConfigMap.Data["ad-guid-migration-status"]
+		if migrationStatus == StatusMigrationFinished {
+			logrus.Debugf("[%v] ad-guid migration has already been completed, refusing to run again at startup", migrateAdUserOperation)
+			return nil
+		}
+	}
+	return UnmigrateAdGUIDUsers(&sc.RESTConfig, false, false)
+}
+
+// UnmigrateAdGUIDUsers will cycle through all users, ctrb, ptrb, tokens and migrate them to an
+// appropriate DN-based PrincipalID.
+func UnmigrateAdGUIDUsers(clientConfig *restclient.Config, dryRun bool, deleteMissingUsers bool) error {
+	if dryRun || os.Getenv("DRY_RUN") == "true" {
+		logrus.Infof("[%v] DRY_RUN is true, no objects will be deleted/modified", migrateAdUserOperation)
+		dryRun = true
+		deleteMissingUsers = false
+	} else if deleteMissingUsers || os.Getenv("AD_DELETE_MISSING_GUID_USERS") == "true" {
+		logrus.Infof("[%v] AD_DELETE_MISSING_GUID_USERS is true, GUID-based users not present in Active Directory will be deleted", migrateAdUserOperation)
+		deleteMissingUsers = true
+	}
+
+	sc, adConfig, err := prepareClientContexts(clientConfig)
+	if err != nil {
+		return err
+	}
+
+	// We'll share this lConn for all lookups to hopefully speed things along
+	lConn, err := ldapConnection(adConfig)
+	if err != nil {
+		return err
+	}
+	defer lConn.Close()
+
+	err = updateMigrationStatus(sc, StatusMigrationField, StatusMigrationRunning)
+	if err != nil {
+		return fmt.Errorf("unable to update migration status configmap: %v", err)
+	}
+
+	users, err := sc.Management.Users("").List(metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("unable to fetch user list: %v", err)
+	}
+
+	usersToMigrate, missingUsers, skippedUsers := identifyMigrationWorkUnits(users, lConn, adConfig)
+	err = collectTokens(&usersToMigrate, sc)
+	if err != nil {
+		return err
+	}
+	err = collectCRTBs(&usersToMigrate, sc)
+	if err != nil {
+		return err
+	}
+	err = collectPRTBs(&usersToMigrate, sc)
+	if err != nil {
+		return err
+	}
+
+	for _, user := range skippedUsers {
+		logrus.Errorf("[%v] unable to migrate user '%v' due to a connection failure; this user will be skipped", migrateAdUserOperation, user.originalUser.Name)
+	}
+	for _, missingUser := range missingUsers {
+		if deleteMissingUsers {
```

Comment:
This should respect the `dryRun` value as well.
```suggestion
		if deleteMissingUsers && !dryRun {
```